File: src/python/xerxes/runtime/change_guard.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Workspace-change guardrails for model-driven edits.

The model can claim a change is safe while the working tree tells a different
story. This module classifies the current Git status and recent tool audit
records so daemon/bridge frontends can surface high-risk edits at turn end.
"""

from __future__ import annotations

import hashlib
import json
import subprocess
from dataclasses import dataclass
from pathlib import Path, PurePosixPath
from typing import Any, Literal

ChangeSeverity = Literal["info", "warning", "error"]

BUILD_CONFIG_PATHS = {
    ".github/workflows/ci.yml",
    ".github/workflows/ci.yaml",
    "Dockerfile",
    "Makefile",
    "pyproject.toml",
    "setup.cfg",
    "setup.py",
    "uv.lock",
}
CRITICAL_SOURCE_PREFIXES = (
    "src/python/xerxes/bridge/",
    "src/python/xerxes/daemon/",
    "src/python/xerxes/runtime/",
    "src/python/xerxes/security/",
    "src/python/xerxes/streaming/",
    "src/python/xerxes/tools/",
)
VERIFICATION_MARKERS = (
    "pytest",
    "ruff check",
    "ruff format",
    "mypy",
    "pre-commit",
    "git diff --check",
    "uv lock --check",
    "docker build",
)
RECENT_TOOL_EXECUTIONS = 50


@dataclass(frozen=True)
class WorkspaceChange:
    """One parsed ``git status --porcelain`` row."""

    status: str
    path: str
    old_path: str = ""

    @property
    def deleted(self) -> bool:
        """Return true when Git marks this path as deleted."""
        return "D" in self.status and not self.untracked

    @property
    def tracked(self) -> bool:
        """Return true for any tracked status row."""
        return self.status != "??"

    @property
    def untracked(self) -> bool:
        """Return true for untracked files."""
        return self.status == "??"


@dataclass(frozen=True)
class ChangeGuardFinding:
    """One risk detected in the working tree."""

    severity: ChangeSeverity
    code: str
    path: str
    message: str


@dataclass(frozen=True)
class ChangeGuardReport:
    """Risk classification plus recent verification evidence."""

    findings: tuple[ChangeGuardFinding, ...]
    verification_commands: tuple[str, ...]
    status_available: bool = True

    @property
    def severity(self) -> ChangeSeverity:
        """Return the highest severity in the report."""
        if any(f.severity == "error" for f in self.findings):
            return "error"
        if any(f.severity == "warning" for f in self.findings):
            return "warning"
        return "info"

    @property
    def should_notify(self) -> bool:
        """Return true when the frontend should show the guard report."""
        if not self.findings:
            return False
        if self.severity == "error":
            return True
        return not self.verification_commands

    @property
    def fingerprint(self) -> str:
        """Stable fingerprint used to suppress duplicate notifications."""
        payload = {
            "findings": [f.__dict__ for f in self.findings],
            "verification_commands": self.verification_commands,
            "status_available": self.status_available,
        }
        encoded = json.dumps(payload, sort_keys=True, ensure_ascii=False).encode("utf-8")
        return hashlib.sha1(encoded).hexdigest()


def analyze_workspace_changes(cwd: Path, tool_executions: list[dict[str, Any]] | None = None) -> ChangeGuardReport:
    """Inspect ``cwd`` with Git and classify risky model-made changes.

    Args:
        cwd: Project directory to inspect.
        tool_executions: Session audit records from recent model tool calls.

    Returns:
        A report suitable for UI notification. Non-Git directories return an
        empty, status-unavailable report instead of raising.
    """
    try:
        proc = subprocess.run(
            ["git", "status", "--porcelain=v1", "--untracked-files=no"],
            cwd=str(cwd),
            capture_output=True,
            text=True,
            check=False,
        )
    except OSError:
        return ChangeGuardReport(findings=(), verification_commands=(), status_available=False)
    if proc.returncode != 0:
        return ChangeGuardReport(findings=(), verification_commands=(), status_available=False)
    return analyze_status_lines(proc.stdout.splitlines(), tool_executions or [])


def analyze_status_lines(
    lines: list[str],
    tool_executions: list[dict[str, Any]] | None = None,
) -> ChangeGuardReport:
    """Classify already-collected porcelain status lines."""
    changes = tuple(parse_porcelain_status(lines))
    findings = tuple(_findings_for_changes(changes))
    verification_commands = tuple(_recent_verification_commands(tool_executions or []))
    return ChangeGuardReport(findings=findings, verification_commands=verification_commands)


def parse_porcelain_status(lines: list[str]) -> list[WorkspaceChange]:
    """Parse ``git status --porcelain=v1`` output."""
    changes: list[WorkspaceChange] = []
    for raw in lines:
        if len(raw) < 4:
            continue
        status = raw[:2]
        path_text = raw[3:]
        old_path = ""
        path = path_text
        if " -> " in path_text:
            old_path, path = path_text.split(" -> ", 1)
        changes.append(
            WorkspaceChange(status=status, path=_normalize_git_path(path), old_path=_normalize_git_path(old_path))
        )
    return changes


def format_change_guard_notification(report: ChangeGuardReport) -> str:
    """Render a concise human-readable notification body."""
    if not report.findings:
        return ""
    lines = ["Risky workspace changes detected:"]
    for finding in report.findings:
        path = f" [{finding.path}]" if finding.path else ""
        lines.append(f"- {finding.message}{path}")
    if report.verification_commands:
        lines.append("")
        lines.append("Recent verification:")
        for command in report.verification_commands[:3]:
            lines.append(f"- {command}")
    else:
        lines.append("")
        lines.append("No recent pytest/ruff/mypy/pre-commit/git diff --check command was found in this session.")
    return "\n".join(lines)


def _findings_for_changes(changes: tuple[WorkspaceChange, ...]) -> list[ChangeGuardFinding]:
    findings: list[ChangeGuardFinding] = []

    deleted_tests = sorted(c.path for c in changes if c.deleted and _is_test_file(c.path))
    if deleted_tests:
        findings.append(
            ChangeGuardFinding(
                severity="error",
                code="deleted-tests",
                path=_sample_paths(deleted_tests),
                message=f"{len(deleted_tests)} tracked test file(s) were deleted",
            )
        )

    deleted_sources = sorted(c.path for c in changes if c.deleted and _is_source_file(c.path))
    if deleted_sources:
        findings.append(
            ChangeGuardFinding(
                severity="warning",
                code="deleted-source",
                path=_sample_paths(deleted_sources),
                message=f"{len(deleted_sources)} source file(s) were deleted",
            )
        )

    build_config = sorted(c.path for c in changes if c.tracked and _is_build_config(c.path))
    if build_config:
        findings.append(
            ChangeGuardFinding(
                severity="warning",
                code="build-config-changed",
                path=_sample_paths(build_config),
                message="build, install, lockfile, or CI configuration changed",
            )
        )

    critical_sources = sorted(c.path for c in changes if c.tracked and _is_critical_source(c.path) and not c.deleted)
    if critical_sources:
        findings.append(
            ChangeGuardFinding(
                severity="warning",
                code="runtime-critical-changed",
                path=_sample_paths(critical_sources),
                message="runtime, daemon, bridge, security, streaming, or tool code changed",
            )
        )

    return findings


def _recent_verification_commands(tool_executions: list[dict[str, Any]]) -> list[str]:
    commands: list[str] = []
    for execution in tool_executions[-RECENT_TOOL_EXECUTIONS:]:
        command = _tool_execution_command(execution)
        if not command:
            continue
        normalized = " ".join(command.split())
        lowered = normalized.lower()
        if any(marker in lowered for marker in VERIFICATION_MARKERS):
            commands.append(_truncate(normalized, 180))
    return commands[-5:]


def _tool_execution_command(execution: dict[str, Any]) -> str:
    name = str(execution.get("name", "")).lower()
    inputs = execution.get("inputs")
    if not isinstance(inputs, dict):
        return ""
    for key in ("cmd", "command", "shell_command"):
        value = inputs.get(key)
        if isinstance(value, str) and value.strip():
            return value
    if "shell" in name or "exec" in name or "bash" in name:
        for value in inputs.values():
            if isinstance(value, str) and value.strip():
                return value
    return ""


def _normalize_git_path(path: str) -> str:
    return path.strip().strip('"').replace("\\", "/")


def _is_test_file(path: str) -> bool:
    pure = PurePosixPath(path)
    return bool(pure.parts and pure.parts[0] == "tests" and pure.name.startswith("test_") and pure.suffix == ".py")


def _is_source_file(path: str) -> bool:
    pure = PurePosixPath(path)
    return bool(path.startswith("src/python/xerxes/") and pure.suffix == ".py")


def _is_build_config(path: str) -> bool:
    return path in BUILD_CONFIG_PATHS or path.startswith(".github/workflows/")


def _is_critical_source(path: str) -> bool:
    return _is_source_file(path) and any(path.startswith(prefix) for prefix in CRITICAL_SOURCE_PREFIXES)


def _sample_paths(paths: list[str], *, limit: int = 5) -> str:
    sample = paths[:limit]
    suffix = f", +{len(paths) - limit} more" if len(paths) > limit else ""
    return ", ".join(sample) + suffix


def _truncate(value: str, limit: int) -> str:
    if len(value) <= limit:
        return value
    return value[: limit - 3] + "..."

File: xerxes/src/runtime/changeGuard.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { createHash } from 'node:crypto'
import { resolve } from 'node:path'

const BUILD_CONFIG_PATHS = new Set([
  '.github/workflows/ci.yml',
  '.github/workflows/ci.yaml',
  '.github/workflows/bun-ci.yml',
  '.github/workflows/bun-release.yml',
  'Dockerfile',
  'Makefile',
  'package.json',
  'bun.lock',
  'bun.lockb',
  'tsconfig.json',
  'xerxes/package.json',
  'xerxes/tsconfig.json',
  'xerxes/tsconfig.ui.json',
  'xerxes/vitest.ui.config.ts',
])

const CRITICAL_SOURCE_PREFIXES = [
  'xerxes/src/bridge/',
  'xerxes/src/daemon/',
  'xerxes/src/runtime/',
  'xerxes/src/security/',
  'xerxes/src/streaming/',
  'xerxes/src/tools/',
] as const

export type ChangeSeverity = 'info' | 'warning' | 'error'

/** Parsed git status porcelain row. */
export class WorkspaceChange {
  readonly oldPath: string
  readonly path: string
  readonly status: string

  constructor(status: string, path: string, oldPath = '') {
    this.status = status
    this.path = path
    this.oldPath = oldPath
    Object.freeze(this)
  }

  get deleted(): boolean {
    return this.status.includes('D') && !this.untracked
  }

  get tracked(): boolean {
    return this.status !== '??'
  }

  get untracked(): boolean {
    return this.status === '??'
  }
}

export interface ChangeGuardFinding {
  readonly code: string
  readonly message: string
  readonly path: string
  readonly severity: ChangeSeverity
}

export interface ChangeGuardReportOptions {
  readonly findings?: readonly ChangeGuardFinding[]
  readonly statusAvailable?: boolean
}

/** Immutable change-risk classification for a working tree. */
export class ChangeGuardReport {
  readonly findings: readonly ChangeGuardFinding[]
  readonly statusAvailable: boolean

  constructor(options: ChangeGuardReportOptions = {}) {
    this.findings = Object.freeze((options.findings ?? []).map(finding => Object.freeze({ ...finding })))
    this.statusAvailable = options.statusAvailable ?? true
    Object.freeze(this)
  }

  get severity(): ChangeSeverity {
    if (this.findings.some(finding => finding.severity === 'error')) return 'error'
    if (this.findings.some(finding => finding.severity === 'warning')) return 'warning'
    return 'info'
  }

  /**
   * Only irreversible-looking damage is worth an unsolicited line in the turn report.
   * Warnings used to fire whenever no verification command had been seen, which turned
   * every routine edit into a chore reminder the model learned to ignore.
   */
  get shouldNotify(): boolean {
    return this.severity === 'error'
  }

  /** Stable SHA-1 fingerprint suitable for duplicate-notification suppression. */
  get fingerprint(): string {
    const payload = JSON.stringify({
      findings: this.findings,
      status_available: this.statusAvailable,
    })
    return createHash('sha1').update(payload, 'utf8').digest('hex')
  }
}

export interface CommandResult {
  readonly exitCode: number
  readonly stdout: string
  readonly stderr?: string
}

export type CommandRunner = (
  args: readonly string[],
  options: { readonly cwd: string },
) => CommandResult | Promise<CommandResult>

export interface AnalyzeWorkspaceChangesOptions {
  readonly commandRunner?: CommandRunner
}

/**
 * Inspect git status using an injectable command runner.
 *
 * A missing Git executable or nonzero status produces an empty,
 * status-unavailable report rather than throwing into the turn loop.
 */
export async function analyzeWorkspaceChanges(
  cwd: string,
  options: AnalyzeWorkspaceChangesOptions = {},
): Promise<ChangeGuardReport> {
  const runner = options.commandRunner ?? defaultCommandRunner
  try {
    const result = await runner(['git', 'status', '--porcelain=v1', '--untracked-files=no'], { cwd })
    if (result.exitCode !== 0) return new ChangeGuardReport({ statusAvailable: false })
    return analyzeStatusLines(result.stdout.split(/\r?\n/))
  } catch {
    return new ChangeGuardReport({ statusAvailable: false })
  }
}

/**
 * Absolute paths that currently differ from HEAD, deletions excluded.
 *
 * Diagnostics need the set of files a turn may have touched; untracked files are
 * included because a freshly created source file is exactly where new errors land.
 * Porcelain paths are repo-root-relative even when git runs in a subdirectory, so
 * they are resolved against the toplevel — resolving against cwd would silently
 * produce paths that match nothing in a nested-package checkout.
 */
export async function changedWorkspacePaths(
  cwd: string,
  options: AnalyzeWorkspaceChangesOptions = {},
): Promise<string[]> {
  const runner = options.commandRunner ?? defaultCommandRunner
  try {
    const root = await runner(['git', 'rev-parse', '--show-toplevel'], { cwd })
    if (root.exitCode !== 0) return []
    const base = root.stdout.trim() || cwd
    const result = await runner(['git', 'status', '--porcelain=v1', '--untracked-files=all'], { cwd })
    if (result.exitCode !== 0) return []
    return parsePorcelainStatus(result.stdout.split(/\r?\n/))
      .filter(change => !change.deleted && change.path)
      .map(change => resolve(base, change.path))
  } catch {
    return []
  }
}

/** Classify already-collected git porcelain lines without shelling out. */
export function analyzeStatusLines(lines: readonly string[]): ChangeGuardReport {
  return new ChangeGuardReport({ findings: findingsForChanges(parsePorcelainStatus(lines)) })
}

/** Parse the portable subset of git status porcelain consumed by the native guard. */
export function parsePorcelainStatus(lines: readonly string[]): WorkspaceChange[] {
  const changes: WorkspaceChange[] = []
  for (const raw of lines) {
    if (raw.length < 4) continue
    const status = raw.slice(0, 2)
    const pathText = raw.slice(3)
    const rename = pathText.indexOf(' -> ')
    const oldPath = rename < 0 ? '' : normalizeGitPath(pathText.slice(0, rename))
    const path = normalizeGitPath(rename < 0 ? pathText : pathText.slice(rename + 4))
    changes.push(new WorkspaceChange(status, path, oldPath))
  }
  return changes
}

/** Render a concise frontend notification for a non-empty risk report. */
export function formatChangeGuardNotification(report: ChangeGuardReport): string {
  if (!report.findings.length) return ''
  const lines = ['Risky workspace changes detected:']
  for (const finding of report.findings) {
    lines.push('- ' + finding.message + (finding.path ? ' [' + finding.path + ']' : ''))
  }
  return lines.join('\n')
}

function findingsForChanges(changes: readonly WorkspaceChange[]): ChangeGuardFinding[] {
  const findings: ChangeGuardFinding[] = []
  const deletedTests = sorted(changes.filter(change => change.deleted && isTestFile(change.path)).map(change => change.path))
  if (deletedTests.length) {
    findings.push(Object.freeze({
      severity: 'error',
      code: 'deleted-tests',
      path: samplePaths(deletedTests),
      message: deletedTests.length + ' tracked test file(s) were deleted',
    }))
  }

  const deletedSources = sorted(changes.filter(change => change.deleted && isSourceFile(change.path)).map(change => change.path))
  if (deletedSources.length) {
    findings.push(Object.freeze({
      severity: 'warning',
      code: 'deleted-source',
      path: samplePaths(deletedSources),
      message: deletedSources.length + ' source file(s) were deleted',
    }))
  }

  const buildConfigs = sorted(changes.filter(change => change.tracked && isBuildConfig(change.path)).map(change => change.path))
  if (buildConfigs.length) {
    findings.push(Object.freeze({
      severity: 'warning',
      code: 'build-config-changed',
      path: samplePaths(buildConfigs),
      message: 'build, install, lockfile, or CI configuration changed',
    }))
  }

  const critical = sorted(changes
    .filter(change => change.tracked && !change.deleted && isCriticalSource(change.path))
    .map(change => change.path))
  if (critical.length) {
    findings.push(Object.freeze({
      severity: 'warning',
      code: 'runtime-critical-changed',
      path: samplePaths(critical),
      message: 'runtime, daemon, bridge, security, streaming, or tool code changed',
    }))
  }
  return findings
}

/**
 * Spawn a command and collect both streams.
 *
 * Both pipes are drained concurrently: a checker that writes a large diagnostic
 * dump to stderr (cargo, ruff) would otherwise fill that pipe and deadlock while
 * this side is still blocked reading stdout.
 */
export async function defaultCommandRunner(
  args: readonly string[],
  options: { readonly cwd: string },
): Promise<CommandResult> {
  const process = Bun.spawn([...args], {
    cwd: options.cwd,
    stdin: 'ignore',
    stdout: 'pipe',
    stderr: 'pipe',
  })
  const [stdout, stderr] = await Promise.all([
    new Response(process.stdout).text(),
    new Response(process.stderr).text(),
  ])
  const exitCode = await process.exited
  return { exitCode, stdout, stderr }
}

function normalizeGitPath(value: string): string {
  return value.trim().replace(/^"|"$/g, '').replaceAll('\\', '/')
}

function isTestFile(path: string): boolean {
  return (path.startsWith('xerxes/test/') && /\.(test|spec)\.ts$/.test(path))
    || (path.startsWith('xerxes/src/ui/') && /\.(test|spec)\.tsx?$/.test(path))
}

function isSourceFile(path: string): boolean {
  return path.startsWith('xerxes/src/') && /\.tsx?$/.test(path)
}

function isBuildConfig(path: string): boolean {
  return BUILD_CONFIG_PATHS.has(path) || path.startsWith('.github/workflows/')
}

function isCriticalSource(path: string): boolean {
  return isSourceFile(path) && CRITICAL_SOURCE_PREFIXES.some(prefix => path.startsWith(prefix))
}

function samplePaths(paths: readonly string[], limit = 5): string {
  const sample = paths.slice(0, limit)
  return sample.join(', ') + (paths.length > limit ? ', +' + (paths.length - limit) + ' more' : '')
}

function sorted(values: readonly string[]): string[] {
  return [...values].sort((left, right) => left.localeCompare(right))
}
